File: services/factory/routers/agents.py
"""agents.py — Spawned agent registry: list, spawn, stop, delete."""
from __future__ import annotations

import os
import time
from pathlib import Path

import docker
import docker.errors
import httpx
from fastapi import APIRouter, HTTPException, Request
from fastapi.responses import StreamingResponse
from pydantic import BaseModel

from database import DB_PATH, get_conn
from routers.events import log_event

router = APIRouter(prefix="/agents", tags=["agents"])

# ── Config ────────────────────────────────────────────────────────────────────
AGENT_IMAGE      = os.environ.get("AGENT_IMAGE",      "nvnnnbt-agent:latest")
COMPOSE_NETWORK  = os.environ.get("COMPOSE_NETWORK",  "nvnnnbt-net")
OLLAMA_URL       = os.environ.get("OLLAMA_BASE_URL",   "http://ollama:11434")
HOST_PROJECT     = os.environ.get("HOST_PROJECT_PATH", "")   # absolute path on HOST machine
AGENT_HOST       = os.environ.get("AGENT_PROXY_HOST", "host.docker.internal")

# Port bands — configurable via env so multiple instances don't collide.
# AGENT_GATEWAY_PORT_BASE: base of 70-port band for gateway (unused band, reserved)
# AGENT_CHAT_PORT_BASE:    base of 70-port band for chat (the port users connect to)
_GATEWAY_BASE = int(os.environ.get("AGENT_GATEWAY_PORT_BASE", "4230"))
_CHAT_BASE    = int(os.environ.get("AGENT_CHAT_PORT_BASE",    "4330"))
_GATEWAY_BAND = range(_GATEWAY_BASE, _GATEWAY_BASE + 70)
_CHAT_BAND    = range(_CHAT_BASE,    _CHAT_BASE    + 70)

# ── Helpers ───────────────────────────────────────────────────────────────────

def _docker_client():
    try:
        return docker.from_env()
    except Exception as e:
        raise HTTPException(status_code=503, detail=f"Docker unavailable: {e}")


def _claim_port(band: range, used: set[int]) -> int:
    for p in band:
        if p not in used:
            return p
    raise HTTPException(status_code=507, detail="No free ports in band")


def _used_ports() -> set[int]:
    with get_conn(DB_PATH) as conn:
        rows = conn.execute("SELECT ports FROM spawned_agents WHERE status = 'running'").fetchall()
    ports: set[int] = set()
    for row in rows:
        if row[0]:
            for v in row[0].split(","):
                v = v.strip()
                if v.isdigit():
                    ports.add(int(v))
    return ports


# ── Helpers: parse ports string ───────────────────────────────────────────────

def _parse_ports(ports_str: str) -> tuple[int | None, int | None]:
    """Return (gateway_port, chat_port) from stored 'gateway,chat' string."""
    if not ports_str:
        return None, None
    parts = [p.strip() for p in ports_str.split(",") if p.strip().isdigit()]
    gateway = int(parts[0]) if len(parts) > 0 else None
    chat    = int(parts[1]) if len(parts) > 1 else None
    return gateway, chat


# ── List agents ───────────────────────────────────────────────────────────────

@router.get("")
async def list_agents() -> dict:
    with get_conn(DB_PATH) as conn:
        rows = conn.execute(
            "SELECT slug, container, ports, model, status, session_type, created_at, workspace_slug "
            "FROM spawned_agents ORDER BY created_at DESC"
        ).fetchall()
    agents = []
    for r in rows:
        gateway_port, chat_port = _parse_ports(r[2])
        agents.append({
            "slug":          r[0],
            "container":     r[1],
            "ports":         r[2],
            "gateway_port":  gateway_port,
            "chat_port":     chat_port,
            "port":          chat_port,   # alias used by UI
            "model":         r[3],
            "status":        r[4],
            "session_type":  r[5],
            "created_at":    r[6],
            "workspace_slug": r[7] or r[0],
        })
    # Sync live container status
    try:
        dc = _docker_client()
        for a in agents:
            if a["container"]:
                try:
                    c = dc.containers.get(a["container"])
                    live = c.status   # "running", "exited", etc.
                    if live != a["status"]:
                        a["status"] = live
                        with get_conn(DB_PATH) as conn:
                            conn.execute(
                                "UPDATE spawned_agents SET status=? WHERE container=?",
                                (live, a["container"]),
                            )
                except docker.errors.NotFound:
                    a["status"] = "missing"
    except HTTPException:
        pass
    return {"agents": agents}


# ── Spawn agent ───────────────────────────────────────────────────────────────

class SpawnBody(BaseModel):
    slug: str
    model: str = "qwen3:1.7b"
    grok_api_key: str = ""


@router.post("")
async def spawn_agent(body: SpawnBody) -> dict:
    if not HOST_PROJECT:
        raise HTTPException(
            status_code=503,
            detail="HOST_PROJECT_PATH env var not set — required for bind-mounting workspace",
        )

    # Validate workspace exists (check via container-internal mount, not host path)
    workspaces_root = os.environ.get("WORKSPACES_ROOT", "/app/data/workspaces")
    ws_path_container = f"{workspaces_root}/{body.slug}"
    if not Path(ws_path_container).exists():
        raise HTTPException(status_code=404, detail=f"Workspace '{body.slug}' not found")
    ws_path_host = str(Path(HOST_PROJECT) / "data" / "workspaces" / body.slug)

    # Check not already running
    with get_conn(DB_PATH) as conn:
        existing = conn.execute(
            "SELECT container FROM spawned_agents WHERE slug=? AND status='running'",
            (body.slug,),
        ).fetchone()
    if existing:
        raise HTTPException(status_code=409, detail=f"Agent '{body.slug}' is already running")

    used = _used_ports()
    gateway_port = _claim_port(_GATEWAY_BAND, used)
    chat_port    = _claim_port(_CHAT_BAND, used | {gateway_port})

    container_name = f"nvnnnbt-agent-{body.slug}"
    dc = _docker_client()

    try:
        container = dc.containers.run(
            AGENT_IMAGE,
            name=container_name,
            detach=True,
            network=COMPOSE_NETWORK,
            ports={
                "6161/tcp": ("0.0.0.0", chat_port),
            },
            environment={
                "WORKSPACE_PATH": ws_path_container,
                "SKILLSETS_PATH": "/app/data/workspaces",
                "AGENT_PORT":     "6161",
                "AGENT_MODE":     "production",
                "OLLAMA_BASE_URL": OLLAMA_URL,
                "GROK_API_KEY":   body.grok_api_key,
            },
            volumes={
                ws_path_host: {"bind": ws_path_container, "mode": "rw"},
                str(Path(HOST_PROJECT) / "data" / "workspaces"): {
                    "bind": "/app/data/workspaces", "mode": "ro",
                },
            },
            restart_policy={"Name": "unless-stopped"},
        )
    except docker.errors.ImageNotFound:
        raise HTTPException(status_code=503, detail=f"Image '{AGENT_IMAGE}' not found")
    except docker.errors.APIError as e:
        raise HTTPException(status_code=500, detail=str(e))

    ports_str = f"{gateway_port},{chat_port}"
    with get_conn(DB_PATH) as conn:
        conn.execute(
            "INSERT OR REPLACE INTO spawned_agents "
            "(slug, container, ports, model, status, session_type, workspace_slug, created_at) "
            "VALUES (?, ?, ?, ?, 'running', 'production', ?, ?)",
            (body.slug, container_name, ports_str, body.model, body.slug, int(time.time())),
        )

    log_event("agent_spawn", target_type="agent", target_slug=body.slug,
              detail=f"Spawned '{body.slug}' on chat_port={chat_port}")
    return {"ok": True, "slug": body.slug, "container": container_name,
            "chat_port": chat_port, "port": chat_port}


# ── Stop agent ────────────────────────────────────────────────────────────────

@router.post("/{slug}/stop")
async def stop_agent(slug: str) -> dict:
    with get_conn(DB_PATH) as conn:
        row = conn.execute(
            "SELECT container FROM spawned_agents WHERE slug=? AND status='running'",
            (slug,),
        ).fetchone()
    if not row:
        raise HTTPException(status_code=404, detail=f"No running agent '{slug}'")

    dc = _docker_client()
    try:
        c = dc.containers.get(row[0])
        c.stop(timeout=10)
    except docker.errors.NotFound:
        pass

    with get_conn(DB_PATH) as conn:
        conn.execute("UPDATE spawned_agents SET status='stopped' WHERE slug=?", (slug,))

    log_event("agent_stop", target_type="agent", target_slug=slug,
              detail=f"Agent '{slug}' stopped")
    return {"ok": True}


# ── Delete agent ──────────────────────────────────────────────────────────────

@router.delete("/{slug}")
async def delete_agent(slug: str) -> dict:
    with get_conn(DB_PATH) as conn:
        row = conn.execute(
            "SELECT container FROM spawned_agents WHERE slug=?", (slug,)
        ).fetchone()
    if not row:
        raise HTTPException(status_code=404, detail=f"No agent record for '{slug}'")

    dc = _docker_client()
    try:
        c = dc.containers.get(row[0])
        c.stop(timeout=5)
        c.remove(force=True)
    except docker.errors.NotFound:
        pass

    with get_conn(DB_PATH) as conn:
        conn.execute("DELETE FROM spawned_agents WHERE slug=?", (slug,))

    log_event("agent_stop", target_type="agent", target_slug=slug,
              detail=f"Agent '{slug}' deleted")
    return {"ok": True}


# ── Start agent ───────────────────────────────────────────────────────────────

@router.post("/{slug}/start")
async def start_agent(slug: str) -> dict:
    with get_conn(DB_PATH) as conn:
        row = conn.execute(
            "SELECT container FROM spawned_agents WHERE slug=?", (slug,)
        ).fetchone()
    if not row:
        raise HTTPException(status_code=404, detail=f"No agent record for '{slug}'")

    dc = _docker_client()
    try:
        c = dc.containers.get(row[0])
        c.start()
    except docker.errors.NotFound:
        raise HTTPException(status_code=404, detail=f"Container for '{slug}' no longer exists — delete and respawn")

    with get_conn(DB_PATH) as conn:
        conn.execute("UPDATE spawned_agents SET status='running' WHERE slug=?", (slug,))

    log_event("agent_spawn", target_type="agent", target_slug=slug,
              detail=f"Agent '{slug}' started")
    return {"ok": True}


# ── Proxy to spawned agent ────────────────────────────────────────────────────
# Routes all requests through factory so the UI never needs a direct port link.
# Reachable via nginx at: /api/factory/agents/{slug}/proxy/{path}

def _agent_base_url(slug: str) -> str:
    """Return http://host:port for a running spawned agent."""
    with get_conn(DB_PATH) as conn:
        row = conn.execute(
            "SELECT ports FROM spawned_agents WHERE slug=? AND status='running'",
            (slug,),
        ).fetchone()
    if not row:
        raise HTTPException(status_code=404, detail=f"Agent '{slug}' not running")
    _, chat_port = _parse_ports(row[0])
    if not chat_port:
        raise HTTPException(status_code=500, detail=f"No chat port recorded for '{slug}'")
    return f"http://{AGENT_HOST}:{chat_port}"


@router.api_route("/{slug}/proxy/{path:path}", methods=["GET", "POST", "PUT", "DELETE"])
async def agent_proxy(slug: str, path: str, request: Request):
    """Transparent proxy — forwards request to spawned agent, streams SSE back."""
    base = _agent_base_url(slug)
    url  = f"{base}/{path}" if path else f"{base}/"

    body    = await request.body()
    headers = {
        k: v for k, v in request.headers.items()
        if k.lower() not in ("host", "content-length")
    }

    # Detect SSE / streaming requests
    wants_stream = (
        request.headers.get("accept", "") == "text/event-stream"
        or path in ("chat", "lab/chat")
    )

    if wants_stream:
        async def _stream():
            async with httpx.AsyncClient(timeout=None) as client:
                async with client.stream(
                    request.method, url,
                    content=body, headers=headers,
                ) as r:
                    async for chunk in r.aiter_bytes():
                        yield chunk
        return StreamingResponse(
            _stream(),
            media_type="text/event-stream",
            headers={"Cache-Control": "no-cache", "X-Accel-Buffering": "no"},
        )

    async with httpx.AsyncClient(timeout=30) as client:
        r = await client.request(
            request.method, url,
            content=body, headers=headers,
        )
    return StreamingResponse(
        iter([r.content]),
        status_code=r.status_code,
        media_type=r.headers.get("content-type", "application/octet-stream"),
    )

File: services/api/src/routes/agents.ts
import { Router } from 'express'
import Dockerode from 'dockerode'
import { createProxyMiddleware } from 'http-proxy-middleware'
import { db } from '../db/database.js'

const router  = Router()
const docker  = new Dockerode({ socketPath: '/var/run/docker.sock' })

// Host-side project root — needed for bind mounts when spawning containers.
// Set HOST_PROJECT_PATH in .env to the absolute path of the nvnNNBT folder on the host.
const HOST_PROJECT_PATH = process.env.HOST_PROJECT_PATH ?? ''
const AGENT_IMAGE       = process.env.AGENT_IMAGE ?? 'nvnnnbt-agent:latest'
const NETWORK_NAME      = process.env.COMPOSE_NETWORK ?? 'nvnnnbt-net'
const OLLAMA_URL        = process.env.OLLAMA_BASE_URL ?? 'http://ollama:11434'

// Port range for spawned agents (host-side)
const PORT_MIN = 6200
const PORT_MAX = 6299

// ── Helpers ───────────────────────────────────────────────────────────────

function usedPorts(): Set<number> {
  const rows = db.prepare('SELECT port FROM agents').all() as { port: number }[]
  return new Set(rows.map(r => r.port))
}

function nextPort(): number {
  const used = usedPorts()
  for (let p = PORT_MIN; p <= PORT_MAX; p++) {
    if (!used.has(p)) return p
  }
  throw new Error('No free ports in range 6200–6299')
}

function slugSafe(s: string): boolean {
  return /^[a-z0-9][a-z0-9-]{0,30}$/.test(s)
}

async function syncStatuses(): Promise<void> {
  const rows = db.prepare("SELECT slug, container FROM agents").all() as { slug: string; container: string }[]
  const live  = await docker.listContainers({ all: true })
  const byName = new Map(live.flatMap(c => c.Names.map(n => [n.replace(/^\//, ''), c])))

  const update = db.prepare("UPDATE agents SET status=? WHERE slug=?")
  for (const row of rows) {
    const c = byName.get(row.container)
    const status = c ? c.State : 'removed'
    update.run(status, row.slug)
  }
}

// ── GET /agents ───────────────────────────────────────────────────────────
router.get('/', async (_req, res) => {
  try {
    await syncStatuses()
    const agents = db.prepare('SELECT * FROM agents ORDER BY created_at DESC').all()
    res.json({ agents })
  } catch (e) {
    res.status(500).json({ error: (e as Error).message })
  }
})

// ── POST /agents — spawn a new agent container ─────────────────────────────
router.post('/', async (req, res) => {
  const { slug, display_name, model } = req.body as {
    slug?: string
    display_name?: string
    model?: string
  }

  if (!slug || !slugSafe(slug)) {
    res.status(400).json({ error: 'slug required — lowercase alphanumeric + hyphens, max 31 chars' })
    return
  }
  if (!display_name?.trim()) {
    res.status(400).json({ error: 'display_name required' })
    return
  }
  if (!HOST_PROJECT_PATH) {
    res.status(500).json({ error: 'HOST_PROJECT_PATH not set — add it to .env' })
    return
  }

  const existing = db.prepare('SELECT slug FROM agents WHERE slug = ?').get(slug)
  if (existing) {
    res.status(409).json({ error: `agent '${slug}' already exists` })
    return
  }

  // Workspace must exist on the host before spawning.
  // If it's missing, Docker will create an empty dir — the agent will start but lack identity files.
  const workspaceHostPath = `${HOST_PROJECT_PATH}/data/workspaces/${slug}`
  const skillsetsHostPath = `${HOST_PROJECT_PATH}/skill-sets`
  const containerName     = `nvnnnbt-agent-${slug}`
  const chosenModel       = (model ?? 'qwen3:1.7b').trim()
  let   port: number

  try {
    port = nextPort()
  } catch (e) {
    res.status(503).json({ error: (e as Error).message })
    return
  }

  try {
    const container = await docker.createContainer({
      name:  containerName,
      Image: AGENT_IMAGE,
      Env: [
        'WORKSPACE_PATH=/workspace',
        'SKILLSETS_PATH=/skill-sets',
        'AGENT_PORT=6161',
        `OLLAMA_BASE_URL=${OLLAMA_URL}`,
      ],
      ExposedPorts: { '6161/tcp': {} },
      HostConfig: {
        Binds: [
          `${workspaceHostPath}:/workspace`,
          `${skillsetsHostPath}:/skill-sets:ro`,
        ],
        PortBindings: {
          '6161/tcp': [{ HostPort: String(port) }],
        },
        RestartPolicy: { Name: 'unless-stopped' },
        NetworkMode: NETWORK_NAME,
      },
    })

    await container.start()

    db.prepare(`
      INSERT INTO agents (slug, display_name, model, port, container, status, started_at)
      VALUES (?, ?, ?, ?, ?, 'running', datetime('now'))
    `).run(slug, display_name.trim(), chosenModel, port, containerName)

    res.status(201).json({ slug, display_name: display_name.trim(), model: chosenModel, port, container: containerName })
  } catch (e) {
    // Clean up if container was created but start failed
    try { docker.getContainer(containerName).remove({ force: true }) } catch { /* ignore */ }
    res.status(500).json({ error: (e as Error).message })
  }
})

// ── GET /agents/:slug/status ──────────────────────────────────────────────
router.get('/:slug/status', async (req, res) => {
  const row = db.prepare('SELECT * FROM agents WHERE slug = ?').get(req.params.slug) as any
  if (!row) { res.status(404).json({ error: 'not found' }); return }

  try {
    const info   = await docker.getContainer(row.container).inspect()
    const status = info.State.Running ? 'running' : info.State.Status
    db.prepare("UPDATE agents SET status=? WHERE slug=?").run(status, row.slug)
    res.json({ slug: row.slug, status, container: row.container, port: row.port })
  } catch (e: any) {
    // 404 from Docker means container was removed outside compose
    if (e?.statusCode === 404) {
      db.prepare("UPDATE agents SET status='removed' WHERE slug=?").run(row.slug)
      res.json({ slug: row.slug, status: 'removed', container: row.container, port: row.port })
    } else {
      res.status(500).json({ error: (e as Error).message })
    }
  }
})

// ── POST /agents/:slug/stop ────────────────────────────────────────────────
router.post('/:slug/stop', async (req, res) => {
  const row = db.prepare('SELECT * FROM agents WHERE slug = ?').get(req.params.slug) as any
  if (!row) { res.status(404).json({ error: 'not found' }); return }

  try {
    await docker.getContainer(row.container).stop()
    db.prepare("UPDATE agents SET status='stopped' WHERE slug=?").run(row.slug)
    res.json({ ok: true })
  } catch (e) {
    res.status(500).json({ error: (e as Error).message })
  }
})

// ── POST /agents/:slug/start ───────────────────────────────────────────────
router.post('/:slug/start', async (req, res) => {
  const row = db.prepare('SELECT * FROM agents WHERE slug = ?').get(req.params.slug) as any
  if (!row) { res.status(404).json({ error: 'not found' }); return }

  try {
    await docker.getContainer(row.container).start()
    db.prepare("UPDATE agents SET status='running', started_at=datetime('now') WHERE slug=?").run(row.slug)
    res.json({ ok: true })
  } catch (e) {
    res.status(500).json({ error: (e as Error).message })
  }
})

// ── DELETE /agents/:slug — stop + remove container + db row ───────────────
router.delete('/:slug', async (req, res) => {
  const row = db.prepare('SELECT * FROM agents WHERE slug = ?').get(req.params.slug) as any
  if (!row) { res.status(404).json({ error: 'not found' }); return }

  try {
    await docker.getContainer(row.container).remove({ force: true })
  } catch {
    // Container may already be gone — continue anyway
  }

  db.prepare('DELETE FROM agents WHERE slug=?').run(row.slug)
  res.json({ ok: true })
})

// ── POST /agents/:slug/chat — SSE proxy to spawned agent ──────────────────
// The spawned container is on the same Docker network as the API,
// reachable by container name.
router.post('/:slug/chat', async (req, res) => {
  const row = db.prepare('SELECT * FROM agents WHERE slug = ?').get(req.params.slug) as any
  if (!row) { res.status(404).json({ error: 'not found' }); return }
  if (row.status !== 'running') { res.status(503).json({ error: `agent '${req.params.slug}' is not running` }); return }

  const target = `http://${row.container}:6161`
  const proxy = createProxyMiddleware({
    target,
    changeOrigin: true,
    selfHandleResponse: false,
  })
  return proxy(req, res, (err?: unknown) => {
    if (err) {
      if (!res.headersSent) {
        res.status(502).json({ error: 'Agent not reachable' })
      }
    }
  })
})

export { router as agentsRouter }
